In [1]:
import pandas as pd
import numpy as np

#### About the dataset:
- The contains details of all matches played by Indian Cricket team of which Sachin was a part of that Game.
- We see that runs by Sachin, no. of fours, no. of sixes, no. of sixes, sr, Inns, Opp teams etc are recorded in the data
- 1 row of the data represents 1 single game played by India for which Sachin was also a part of.

In [2]:
data = pd.read_csv('Sachin_ODI.csv')
data.head()

runs  NotOut mins  bf  fours  sixes      sr  Inns           Opp    Ground  \
0    13       0   30  15      3      0   86.66     1   New Zealand    Napier   
1    37       0   75  51      3      1   72.54     2  South Africa  Hamilton   
2    47       0   65  40      7      0  117.50     2     Australia   Dunedin   
3    48       0   37  30      9      1  160.00     2    Bangladesh   Sharjah   
4     4       0   13   9      1      0   44.44     2      Pakistan   Sharjah   

         Date        Winner    Won  century  
0  1995-02-16   New Zealand  False    False  
1  1995-02-18  South Africa  False    False  
2  1995-02-22         India   True    False  
3  1995-04-05         India   True    False  
4  1995-04-07      Pakistan  False    False

In [3]:
data.shape

(360, 14)

#### Total matches played by India for which Sachin was also a part of is 360 ODI matches.

## How to get the Sample Space for the data?

In [4]:
data.shape[0]

360

In [5]:
len(data)

360

- The total Sample Space for the data is 360.

## Event A : India winning the match

In [6]:
## For Event A, Find out the sample space

data['Won'].value_counts()

True     184
False    176
Name: Won, dtype: int64

- The Sample Space for India winning the match i.e. Event A is 184

### Probability of Event A:

In [9]:
## Static Approach to find the probability of an Event

p_a = 184 / 360
p_a

0.5111111111111111

In [11]:
### Pandas Approach to find the probability of an Event A

data_won = data[data['Won'] == True]
data_won.head()

runs  NotOut mins   bf  fours  sixes      sr  Inns          Opp    Ground  \
2    47       0   65   40      7      0  117.50     2    Australia   Dunedin   
3    48       0   37   30      9      1  160.00     2   Bangladesh   Sharjah   
5   112       1  137  107     15      1  104.67     2    Sri Lanka   Sharjah   
6    41       0   51   41      5      0  100.00     2    Sri Lanka   Sharjah   
8    39       0   79   51      4      0   76.47     2  New Zealand  Amritsar   

         Date Winner   Won  century  
2  1995-02-22  India  True    False  
3  1995-04-05  India  True    False  
5  1995-04-09  India  True     True  
6  1995-04-14  India  True    False  
8  1995-11-18  India  True    False

In [12]:
data_won.shape[0]

184

In [13]:
data.shape[0]

360

In [15]:
p_a = data_won.shape[0] / data.shape[0]
p_a

0.5111111111111111

## Event B : India losing the match

In [16]:
## Static appraoch
176 / 360

0.4888888888888889

In [17]:
## Pandas Approach for Event B

data_lose = data[data['Won'] == False]
data_lose.head()

runs  NotOut mins  bf  fours  sixes      sr  Inns           Opp  \
0     13       0   30  15      3      0   86.66     1   New Zealand   
1     37       0   75  51      3      1   72.54     2  South Africa   
4      4       0   13   9      1      0   44.44     2      Pakistan   
7     30       0   30  20      5      0  150.00     1   New Zealand   
10    65       0   96  59      9      1  110.16     2   New Zealand   

        Ground        Date        Winner    Won  century  
0       Napier  1995-02-16   New Zealand  False    False  
1     Hamilton  1995-02-18  South Africa  False    False  
4      Sharjah  1995-04-07      Pakistan  False    False  
7   Jamshedpur  1995-11-15   New Zealand  False    False  
10      Nagpur  1995-11-26   New Zealand  False    False

In [18]:
data_lose.shape[0]

176

In [19]:
data.shape[0]

360

In [21]:
p_b = data_lose.shape[0] / data.shape[0]
p_b

0.4888888888888889

In [23]:
## Probability Rules : Event B is just the complement of Event A

p_a_c = 1 - p_a
p_a_c

0.48888888888888893

## Observations :

- India has chances of winning 51% times
- India has chances of losing 48% times 

## Event C : Sachin scoring a Century

In [25]:
### Find out the Sample Space for Event C

data['century'].value_counts()

False    314
True      46
Name: century, dtype: int64

## Probability of Sachin scoring a Century

In [26]:
## Static approach

46 / 360

0.12777777777777777

### Sachin has a chance of scoring a century of only 12.8%

In [28]:
## Pandas Approach

data_cen = data[data['century'] == True]
data_cen

runs  NotOut mins   bf  fours  sixes      sr  Inns           Opp  \
5     112       1  137  107     15      1  104.67     2     Sri Lanka   
12    127       1  175  138     15      1   92.02     2         Kenya   
15    137       0  198  137      8      5  100.00     1     Sri Lanka   
20    100       0    -  111      9      1   90.09     1      Pakistan   
23    118       0    -  140      8      2   84.28     1      Pakistan   
27    110       0    -  138      5      1   79.71     1     Sri Lanka   
41    114       0  157  126     14      0   90.47     1  South Africa   
47    104       0  121   97      8      1  107.21     2      Zimbabwe   
55    117       0  188  137     13      2   85.40     2   New Zealand   
85    100       0  132   89      5      7  112.35     2     Australia   
91    143       0  187  131      9      5  109.16     2     Australia   
92    134       0  195  131     12      3  102.29     2     Australia   
95    100       1  156  103     13      0   97.08     2         Kenya   
99    128       0  182  131      8      2   97.70     1     Sri Lanka   
101   127       1  155  130     13      1   97.69     2      Zimbabwe   
104   141       0  191  128     13      3  110.15     1     Australia   
107   118       1  173  112     14      2  105.35     2      Zimbabwe   
110   124       1  140   92     12      6  134.78     2      Zimbabwe   
116   140       1  114  101     16      3  138.61     1         Kenya   
123   120       0  191  141     11      2   85.10     1     Sri Lanka   
128   186       1  221  150     20      3  124.00     1   New Zealand   
143   122       0  195  138     12      0   88.40     2  South Africa   
155   101       0    -  140      3      1   72.14     1     Sri Lanka   
162   146       0  200  153     15      2   95.42     1      Zimbabwe   
167   139       0  197  125     19      0  111.20     1     Australia   
173   122       1  210  131     12      1   93.12     2   West Indies   
175   101       0  209  129      9      0   78.29     1  South Africa   
179   146       0  186  132     17      0  110.60     1         Kenya   
191   105       1  167  108      8      1   97.22     1       England   
194   113       0  141  102     12      1  110.78     1     Sri Lanka   
206   152       0  193  151     18      0  100.66     1       Namibia   
215   100       0  158  119      9      1   84.03     1     Australia   
219   102       0  124   91     12      1  112.08     1   New Zealand   
229   141       0  173  135     17      1  104.44     2      Pakistan   
245   123       0  192  130     12      2   94.61     1      Pakistan   
258   100       0  188  113     10      1   88.49     1      Pakistan   
262   141       1    -  148     13      5   95.27     1   West Indies   
276   100       1  110   76     10      1  131.57     1   West Indies   
313   117       1  235  120     10      0   97.50     2     Australia   
322   163       1  206  133     16      5  122.55     1   New Zealand   
325   138       0  214  133     10      1  103.75     1     Sri Lanka   
331   175       0  210  141     19      4  124.11     2     Australia   
338   200       1  226  147     25      3  136.05     1  South Africa   
342   120       0  170  115     10      5  104.34     1       England   
345   111       0  184  101      8      3  109.90     1  South Africa   
358   114       0  205  147     12      1   77.55     1    Bangladesh   

                 Ground        Date        Winner    Won  century  
5               Sharjah  1995-04-09         India   True     True  
12              Cuttack  1996-02-18         India   True     True  
15                Delhi  1996-03-02     Sri Lanka  False     True  
20            Singapore  1996-04-05      Pakistan  False     True  
23              Sharjah  1996-04-15         India   True     True  
27        Colombo (RPS)  1996-08-28     Sri Lanka  False     True  
41               Mumbai  1996-12-14         India   True     True  
47               Benoni  1997

In [29]:
data_cen.shape[0]

46

In [30]:
data.shape[0]

360

In [32]:
p_c = data_cen.shape[0] / data.shape[0]
p_c

0.12777777777777777

## Probability of Sachin not scoring a century

In [33]:
p_c_c = 1 - p_c
p_c_c

0.8722222222222222

## Event D : Probability of India Winning and Sachin Scoring Century

In [37]:
## Find out the Sample Space for the Event D

data_iwsc = data[(data['Won'] == True) & (data['century'] == True)]
data_iwsc

runs  NotOut mins   bf  fours  sixes      sr  Inns           Opp  \
5     112       1  137  107     15      1  104.67     2     Sri Lanka   
12    127       1  175  138     15      1   92.02     2         Kenya   
23    118       0    -  140      8      2   84.28     1      Pakistan   
41    114       0  157  126     14      0   90.47     1  South Africa   
47    104       0  121   97      8      1  107.21     2      Zimbabwe   
55    117       0  188  137     13      2   85.40     2   New Zealand   
85    100       0  132   89      5      7  112.35     2     Australia   
92    134       0  195  131     12      3  102.29     2     Australia   
95    100       1  156  103     13      0   97.08     2         Kenya   
99    128       0  182  131      8      2   97.70     1     Sri Lanka   
101   127       1  155  130     13      1   97.69     2      Zimbabwe   
104   141       0  191  128     13      3  110.15     1     Australia   
107   118       1  173  112     14      2  105.35     2      Zimbabwe   
110   124       1  140   92     12      6  134.78     2      Zimbabwe   
116   140       1  114  101     16      3  138.61     1         Kenya   
123   120       0  191  141     11      2   85.10     1     Sri Lanka   
128   186       1  221  150     20      3  124.00     1   New Zealand   
143   122       0  195  138     12      0   88.40     2  South Africa   
167   139       0  197  125     19      0  111.20     1     Australia   
173   122       1  210  131     12      1   93.12     2   West Indies   
179   146       0  186  132     17      0  110.60     1         Kenya   
194   113       0  141  102     12      1  110.78     1     Sri Lanka   
206   152       0  193  151     18      0  100.66     1       Namibia   
215   100       0  158  119      9      1   84.03     1     Australia   
219   102       0  124   91     12      1  112.08     1   New Zealand   
276   100       1  110   76     10      1  131.57     1   West Indies   
313   117       1  235  120     10      0   97.50     2     Australia   
322   163       1  206  133     16      5  122.55     1   New Zealand   
325   138       0  214  133     10      1  103.75     1     Sri Lanka   
338   200       1  226  147     25      3  136.05     1  South Africa   

                 Ground        Date Winner   Won  century  
5               Sharjah  1995-04-09  India  True     True  
12              Cuttack  1996-02-18  India  True     True  
23              Sharjah  1996-04-15  India  True     True  
41               Mumbai  1996-12-14  India  True     True  
47               Benoni  1997-02-09  India  True     True  
55            Bengaluru  1997-05-14  India  True     True  
85               Kanpur  1998-04-07  India  True     True  
92              Sharjah  1998-04-24  India  True     True  
95              Kolkata  1998-05-31  India  True     True  
99        Colombo (RPS)  1998-07-07  India  True     True  
101            Bulawayo  1998-09-26  India  True     True  
104               Dhaka  1998-10-28  India  True     True  
107             Sharjah  1998-11-08  India  True     True  
110             Sharjah  1998-11-13  India  True     True  
116             Bristol  1999-05-23  India  True     True  
123       Colombo (SSC)  1999-08-29  India  True     True  
128  Hyderabad (Deccan)  1999-11-08  India  True     True  
143            Vadodara  2000-03-17  India  True     True  
167              Indore  2001-03-31  India  True     True  
173              Harare  2001-07-04  India  True     True  
179               Paarl  2001-10-24  India  True     True  
194             Bristol  2002-07-11  India  True     True  
206    Pietermaritzburg  2003-02-23  India  True     True  
215             Gwalior  2003-10-26  India  True     True  
219  Hyderabad (Deccan)  2003-11-15  India  True     True  
276            Vadodara  2007-01-31  India  True     True  
313              Sydney  2008-03-02  India  True     True  
322        Christchurch  2009-03-08  India  True     True  
3

In [38]:
data_iwsc.shape[0]

30

In [39]:
data.shape[0]

360

In [41]:
p_iwsc = data_iwsc.shape[0] / data.shape[0]
p_iwsc

0.08333333333333333

## Observations:

- There is only 8% chances that India will win the match if Sachin is scoring a century in that match
- There is only 8% chance that Sachin might have scored a century if India has won the match

## Simpler way to calculate probabilities : Probability of India Winning and Sachin scoring a century

In [43]:
# pd.crosstab(index = data['Won'], columns = data['century'])
pd.crosstab(index = data['century'], columns = data['Won'], margins = True)

Won      False  True  All
century                  
False      160   154  314
True        16    30   46
All        176   184  360

In [44]:
30 / 360

0.08333333333333333

## Find the probability of India Winning given that Sachin has already scored a century

A : Sachin scoring a century  
B : India winning the match  

P(B|A) i.e. P(India winning the match | Sachin has already scored a century)


In [45]:
# pd.crosstab(index = data['Won'], columns = data['century'])
pd.crosstab(index = data['century'], columns = data['Won'], margins = True)

Won      False  True  All
century                  
False      160   154  314
True        16    30   46
All        176   184  360

- Sample Space of Sachin already scoring a century = 46
- Sample Space of India winning given that Sachin already scored a century = 30

In [46]:
30 / 46

0.6521739130434783

## Observations:

- India has a 65% chance of winning the game if Sachin has already scored a century

In [51]:
### Pandas Approach 

data_cen = data[data['century'] == True]
data_cen.head()

runs  NotOut mins   bf  fours  sixes      sr  Inns        Opp     Ground  \
5    112       1  137  107     15      1  104.67     2  Sri Lanka    Sharjah   
12   127       1  175  138     15      1   92.02     2      Kenya    Cuttack   
15   137       0  198  137      8      5  100.00     1  Sri Lanka      Delhi   
20   100       0    -  111      9      1   90.09     1   Pakistan  Singapore   
23   118       0    -  140      8      2   84.28     1   Pakistan    Sharjah   

          Date     Winner    Won  century  
5   1995-04-09      India   True     True  
12  1996-02-18      India   True     True  
15  1996-03-02  Sri Lanka  False     True  
20  1996-04-05   Pakistan  False     True  
23  1996-04-15      India   True     True

In [53]:
data_cen.shape[0]

46

In [56]:
data_cen_won = data_cen[data_cen['Won'] == True]
data_cen_won.head()

runs  NotOut mins   bf  fours  sixes      sr  Inns           Opp   Ground  \
5    112       1  137  107     15      1  104.67     2     Sri Lanka  Sharjah   
12   127       1  175  138     15      1   92.02     2         Kenya  Cuttack   
23   118       0    -  140      8      2   84.28     1      Pakistan  Sharjah   
41   114       0  157  126     14      0   90.47     1  South Africa   Mumbai   
47   104       0  121   97      8      1  107.21     2      Zimbabwe   Benoni   

          Date Winner   Won  century  
5   1995-04-09  India  True     True  
12  1996-02-18  India  True     True  
23  1996-04-15  India  True     True  
41  1996-12-14  India  True     True  
47  1997-02-09  India  True     True

In [58]:
data_cen_won.shape[0]

30

In [59]:
data_cen_won.shape[0] / data_cen.shape[0]

0.6521739130434783

## Naive Bayes Algorithm:

- The foundation principle on which Naive Bayes is dependent upon is Conditional Probability


#### What is Naive Bayes?
- Naive is classification and Regression Algorithm
- Widely used for classification
- Widely used when the data is very big
- Three types in Naive Bayes - Guassian NB, Multinomial NB, Bernoulli NB